memq-client/src/main/java/com/pinterest/memq/client/commons2/MemqCommonClient.java

```diff
@@ -349,14 +361,19 @@ protected List<Endpoint> getEndpointsToTry() {
    */
   protected void maybeRegisterWriteEndpoint(Endpoint endpoint, String topic) {
     failureCounts.remove(endpoint);
-    if (writeEndpoints.size() < numWriteEndpoints && !writeEndpoints.contains(endpoint)) {
+    List<Endpoint> currentWrites = this.writeEndpoints;
+    if (currentWrites.size() < numWriteEndpoints && !currentWrites.contains(endpoint)) {
       logger.info("Registering write endpoint: " + endpoint + " for topic: " + topic);
-      writeEndpoints.add(endpoint);
+      List<Endpoint> newWrites = new ArrayList<>(currentWrites);
+      newWrites.add(endpoint);
+      this.writeEndpoints = Collections.unmodifiableList(newWrites);
     }
```

Comment:
Is it still possible for two threads to overwrite each other in this `if` block?